### PR TITLE
EL-3962 - AngularJS Add Schematics

### PR DESCRIPTION
**Pull request by ashley-hunter**
#### Checklist

* [x] The contents of this PR are mine to submit. No third party code or other material is being committed.
* [x] New third party dependencies (if any) are linked via `package.json`. Third party dependencies are licenced under one of the following: Apache, MIT, BSD, Mozilla Public License, Eclipse Public License, or Oracle Binary Code License.
* [x] The code in this PR does not contain export-restricted encryption algorithms.

#### Ticket / Issue

https://portal.digitalsafe.net/browse/EL-3962

#### Description of Proposed Changes

- Moving schematics from separate repo into this repo as an add schematic
- Updating getting started page

#### Documentation CI URL

https://pages.github.houston.softwaregrp.net/sepg-docs-qa/UXAspects_CI_UXAspects-ng1_EL-3962-Schematics

### CODE REVIEW

**Comment by ashley-hunter** [comment]
The review is partially blocked as all builds seem to be failing at the minute. Its not able to upload the package to artifactory. The build team are aware of this issue as Tara reported it this morning.

File: schematics/ng-generate/angularjs-hybrid/index.ts [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+}
+
+function addStyles(options: Schema): Rule {
+    return (tree: Tree, _context: SchematicContext) => {
+        const workspace = getWorkspace(tree);
+        const project = getProjectFromWorkspace(workspace, options.project);
+
+        addStyleImportsAfter(
+            project,
+            'node_modules/@ux-aspects/ux-aspects/styles/ux-aspects.css',
```
Comment (donna-marijne) [inline schematics/ng-generate/angularjs-hybrid/index.ts:95]: Because of the hpe-icon conflict issue we had recently, the schematic should replace the style import
```
node_modules/@ux-aspects/ux-aspects/styles/ux-aspects.css
```
with
```
node_modules/@ux-aspects/ux-aspects/styles/ux-aspects-no-legacy.css
```

(The ng1 stylesheet should still be added following the ux-aspects-no-legacy stylesheet.)

File: docs/app/data/licenses.json [hunk 2]
```diff
@@ -2,161 +2,161 @@
...
...\r\nAny arrangements, understandings, or agreements regarding the Licensed Material not stated herein are separate from and independent of the terms and conditions of this Public License.\r\nSection 8 - Interpretation.\r\n\r\nFor the avoidance of doubt, this Public License does not, and shall not be interpreted to, reduce, limit, restrict, or impose conditions on any use of the Licensed Material that could lawfully be made without permission under this Public License.\r\nTo the extent possible, if any provision of this Public License is deemed unenforceable, it shall be automatically reformed to the minimum extent necessary to make it enforceable. If the provision cannot be reformed, it shall be severed from this Public License without affecting the enforceability of the remaining terms and conditions.\r\nNo term or condition of this Public License will be waived and no failure to comply consented to unless expressly agreed to by the Licensor.\r\nNothing in this Public License constitutes or may be interpreted as a limitation upon, or waiver of, any privileges and immunities that apply to the Licensor or You, including from the legal processes of any jurisdiction or authority.\r\n"
```
Comment (donna-marijne) [inline docs/app/data/licenses.json]: Were these changes intentional?

File: docs/app/pages/getting-started/getting-started.component.html [hunk 3]
```diff
@@ -1,6 +1,21 @@
 <uxd-page-header header="Getting Started"></uxd-page-header>
 <uxd-text-page-layout>
-    <h3>Installation</h3>
+
+    <h3 id="add-to-an-existing-project">Add to an Existing Project</h3>
+
+    <p>
+        To add UX Aspects to your Angular CLI or Nx workspace, run the following command in the project root:
```
Comment (donna-marijne) [inline docs/app/pages/getting-started/getting-started.component.html]: I would prefer to always refer to the library as "UX Aspects (AngularJS)" here, to minimize any potential confusion about what these instructions are referring to.

File: Gruntfile.js [hunk 4]
```diff
@@ -54,6 +57,7 @@ module.exports = function(grunt) {
         'styles',
         'minify',
         'assets:library',
+        'schematics',
```
Comment (donna-marijne) [inline Gruntfile.js:61]: I noticed that the schematics were not included in the Artifactory package. There is a separate, quite similar grunt task named "compile" which is run when building both library and docs (npm run build). This would need to be added to that task as well.

**Review by donna-marijne — CHANGES_REQUESTED** [review]
Thanks for updating this! It looks good, just a couple of corrections to make.

File: docs/app/data/licenses.json (hunk 2 above)
Comment (ashley-hunter) [reply docs/app/data/licenses.json]: Hmm, nope, must have been a grunt task perhaps? Anyway, I will revert the changes.

**Comment by ashley-hunter** [comment]
Thanks for the suggestions, they should now be fixed, thanks

**Comment by ashley-hunter** [comment]
I'm guessing the build is failing due to this:

image

**Comment by donna-marijne** [comment]
Looks like it, I've got a failing feature branch as well, but some other builds are passing.

File: schematics/ng-generate/angularjs-hybrid/index.ts [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+            'node_modules/@ux-aspects/ux-aspects-ng1/styles/ux-aspects.css'
+        );
+
+        // use the UX Aspects no legacy styles instead of the default stylesheet
+        replaceImport(
```
Comment (donna-marijne) [inline schematics/ng-generate/angularjs-hybrid/index.ts:91]: I would suggest doing this before the `addStyleImportsAfter` call and then you can safely add the ng1 stylesheet after `node_modules/@ux-aspects/ux-aspects/styles/ux-aspects-no-legacy.css`. This would cover the edge case where someone has manually switched to the "no-legacy" stylesheet - Jane has done this in at least one of her projects (although it's very unlikely that those projects will adopt schematics).

File: schematics/ng-generate/angularjs-hybrid/index.ts [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+    if (method) {
+        // Insert the content at the top of the existing method body
+        change = new InsertChange(source.fileName, method.body!.getStart() + 1, body);
+    } else {
+        // Create the new function at the end of the class
+        change = new InsertChange(
+            source.fileName,
+            klass.members.end,
+            `
+  ${ methodName }() {
```
Comment (donna-marijne) [inline schematics/ng-generate/angularjs-hybrid/index.ts]: I noticed that this causes a linting error due to not having a return type:
```
ERROR: 27:3  typedef  expected call-signature: 'ngDoBootstrap' to have a typedef
```

File: docs/app/pages/getting-started/getting-started.component.html [hunk 7]
```diff
@@ -1,6 +1,21 @@
...
-    <h3>Installation</h3>
+
+    <h3 id="add-to-an-existing-project">Add to an Existing Project</h3>
+
+    <p>
+        To add UX Aspects (AngularJS) to your Angular CLI or Nx workspace, run the following command in the project root:
+    </p>
+
+    <pre>
+ng add @ux-aspects/ux-aspects-ng1 --registry https://svsartifactory.swinfra.net/artifactory/api/npm/company-central-npm/</pre>
```
Comment (donna-marijne) [inline docs/app/pages/getting-started/getting-started.component.html]: We actually don't need `--registry` in this case since we are only deploying open source components.

**Review by donna-marijne — CHANGES_REQUESTED** [review]
Thanks for the update! It works really well except for one (pre-existing) lint issue.

**Comment by ashley-hunter** [comment]
Thanks for the feedback Donna, I have pushed fixes for those comments.

File: schematics/ng-generate/angularjs-hybrid/index.ts [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+            project,
+            'build',
+            'styles',
+            'node_modules/@ux-aspects/ux-aspects/styles/ux-aspects.css',
+            'node_modules/@ux-aspects/ux-aspects/styles/ux-aspects-no-legacy.css');
+
+        addStyleImportsAfter(
+            project,
+            'node_modules/@ux-aspects/ux-aspects/styles/ux-aspects.css',
+            'node_modules/@ux-aspects/ux-aspects-ng1/styles/ux-aspects-no-legacy.css'
```
Comment (donna-marijne) [inline schematics/ng-generate/angularjs-hybrid/index.ts]: These lines just got a bit mixed up, `no-legacy` should be on the previous line, i.e.
```
node_modules/@ux-aspects/ux-aspects/styles/ux-aspects-no-legacy.css
```

**Review by donna-marijne — CHANGES_REQUESTED** [review]

File: schematics/ng-generate/angularjs-hybrid/index.ts (hunk 1 above)
Comment (ashley-hunter) [reply schematics/ng-generate/angularjs-hybrid/index.ts:100]: Whoops, well spotted! Thank you! That should now be fixed

File: schematics/ng-generate/angularjs-hybrid/index.ts (hunk 8 above)
Comment (ashley-hunter) [reply schematics/ng-generate/angularjs-hybrid/index.ts]: Whoops, well spotted! Thank you! That should now be fixed

File: schematics/ng-generate/angularjs-hybrid/index.ts [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+        replaceImport(
+            project,
+            'build',
+            'styles',
+            'node_modules/@ux-aspects/ux-aspects/styles/ux-aspects.css',
+            'node_modules/@ux-aspects/ux-aspects/styles/ux-aspects-no-legacy.css');
+
+        addStyleImportsAfter(
+            project,
+            'node_modules/@ux-aspects/ux-aspects/styles/ux-aspects.css',
```
Comment (donna-marijne) [inline schematics/ng-generate/angularjs-hybrid/index.ts]: This line should be
```
node_modules/@ux-aspects/ux-aspects/styles/ux-aspects-no-legacy.css
```

The ng1 stylesheet needs to have precedence over the standard (no-legacy) stylesheet in case of any conflicts.

**Review by donna-marijne — CHANGES_REQUESTED** [review]

**Review by donna-marijne — APPROVED** [review]
Thank you for making the additional changes! Looks good to me.
